lib/operations/execute_operation_v2.js

```diff
@@ -76,11 +81,133 @@ function executeOperationV2(topology, operation, callback) {
+      } else {
+        return operation.execute(handler);
+      }
     } catch (e) {
       handler(e);
     }
   });
 }
 
+const MongoNetworkError = require('mongodb-core').MongoNetworkError;
+const RETRYABLE_WIRE_VERSION = 6;
+
+// see: https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#terms
+const RETRYABLE_ERROR_CODES = new Set([
+  6, // HostUnreachable
+  7, // HostNotFound
+  89, // NetworkTimeout
+  91, // ShutdownInProgress
+  189, // PrimarySteppedDown
+  9001, // SocketException
+  10107, // NotMaster
+  11600, // InterruptedAtShutdown
+  11602, // InterruptedDueToReplStateChange
+  13435, // NotMasterNoSlaveOk
+  13436 // NotMasterOrSecondary
+]);
+
+/**
+ * Determines whether an error is something the driver should attempt to retry
+ *
+ * @param {MongoError|Error} error
+ */
+function isRetryableError(error) {
+  return (
+    RETRYABLE_ERROR_CODES.has(error.code) ||
+    error instanceof MongoNetworkError ||
+    error.message.match(/not master/) ||
+    error.message.match(/node is recovering/)
+  );
+}
+
+function isSingleServer(topology) {
+  if (topology.type && topology.type === 'server') {
+    return true;
+  }
+
+  if (topology.description && topology.description.type === 'Single') {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Determines whether the provided topology supports retryable writes
+ *
+ * @param {*} topology
+ */
+const isRetryabilitySupported = function(topology) {
+  if (isSingleServer(topology)) {
+    return false;
+  }
+
+  const maxWireVersion = topology.lastIsMaster().maxWireVersion;
+  if (maxWireVersion < RETRYABLE_WIRE_VERSION) {
+    return false;
+  }
+
+  if (!topology.logicalSessionTimeoutMinutes) {
+    return false;
+  }
+
+  return true;
+};
+
+function executeWithServerSelection(topology, operation, callback) {
+  const readPreference = operation.readPreference || ReadPreference.primary;
+  const session = operation.session;
+  const options = Object.assign({}, operation.options);
+
+  const willRetryWrite =
+    operation.hasAspect(Aspect.RETRY) &&
+    operation.canRetry &&
+    session &&
+    isRetryabilitySupported(topology) &&
+    !options.session.inTransaction();
+
+  function callbackWithRetry(err, result) {
+    if (err == null) {
+      return callback(null, result);
+    }
+
+    if (!isRetryableError(err)) {
+      return callback(err);
+    }
+
+    // select a new server, and attempt to retry the operation
+    topology.selectServer(readPreference, (err, server) => {
+      if (err) {
+        callback(err, null);
+        return;
+      }
+
+      operation.execute(server, callback);
+    });
+  }
+
+  // select a server, and execute the operation against it
+  topology.selectServer(readPreference, (err, server) => {
+    if (err) {
+      callback(err, null);
+      return;
+    }
+
+    if (willRetryWrite) {
+      // TODO: I hate this
+      operation.options.willRetryWrite = true;
```

Comment:
I don't think I'm understanding what's going on here. Why do we need to set this on the `operation`?

Reply:
It's a hack. It signals to the wire protocol to decorate the command with a `txnNumber`. See [applySession](https://github.com/mongodb-js/mongodb-core/blob/master/lib/sessions.js#L633) and [command](https://github.com/mongodb-js/mongodb-core/blob/master/lib/wireprotocol/command.js#L36).

It was the minimum I needed to do to get this working, and I am currently trying to brainstorm a way to do this in a more elegant manner. However, it is escaping me at the moment :-(.

Reply:
Oh, I see. I'll also think of some ideas.